File: ffupdater/src/main/java/de/marmaro/krt/ffupdater/app/impl/Iceraven.java
package de.marmaro.krt.ffupdater.app.impl;

import android.content.Context;
import android.os.Build;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import de.marmaro.krt.ffupdater.R;
import de.marmaro.krt.ffupdater.app.impl.fetch.ApiConsumer;
import de.marmaro.krt.ffupdater.app.impl.fetch.github.GithubConsumer;
import de.marmaro.krt.ffupdater.app.impl.fetch.github.dao.Asset;
import de.marmaro.krt.ffupdater.app.impl.fetch.github.dao.Result;
import de.marmaro.krt.ffupdater.app.BaseApp;
import de.marmaro.krt.ffupdater.app.UpdateCheckResult;
import de.marmaro.krt.ffupdater.device.ABI;
import de.marmaro.krt.ffupdater.utils.ParamRuntimeException;

import static de.marmaro.krt.ffupdater.app.UpdateCheckResult.FILE_SIZE_BYTES;
import static de.marmaro.krt.ffupdater.device.ABI.AARCH64;
import static de.marmaro.krt.ffupdater.device.ABI.ARM;
import static de.marmaro.krt.ffupdater.device.ABI.X86;
import static de.marmaro.krt.ffupdater.device.ABI.X86_64;

/**
 * https://api.github.com/repos/fork-maintainers/iceraven-browser/releases
 */
public class Iceraven extends BaseApp {

    @Override
    public String getPackageName() {
        return "io.github.forkmaintainers.iceraven";
    }

    @Override
    public String getDisplayTitle(Context context) {
        return context.getString(R.string.iceraven_title);
    }

    @Override
    public String getDisplayDescription(Context context) {
        return context.getString(R.string.iceraven_description);
    }

    @Override
    public Optional<String> getDisplayWarning(Context context) {
        //TODO
        return Optional.empty();
    }

    @Override
    public String getSignatureHashAsString() {
        return "9c0d22379f487b70a4f9f8bec0173cf91a1644f08f93385b5b782ce37660ba81";
    }

    @Override
    public String getDisplayDownloadSource(Context context) {
        return context.getString(R.string.github);
    }

    @Override
    public Optional<String> getDisplayInstalledVersion(Context context) {
        return getInstalledVersionFromPackageManager(context);
    }

    @Override
    public Optional<String> getInstalledVersion(Context context) {
        return getInstalledVersionFromPackageManager(context);
    }

    @Override
    public int getMinApiLevel() {
        return Build.VERSION_CODES.LOLLIPOP;
    }

    @Override
    public List<ABI> getSupportedAbi() {
        return Arrays.asList(AARCH64, ARM, X86_64, X86);
    }

    @Override
    public UpdateCheckResult updateCheckBlocking(Context context, ABI abi) {
        final Result result = new GithubConsumer.Builder()
                .setApiConsumer(new ApiConsumer())
                .setRepoOwner("fork-maintainers")
                .setRepoName("iceraven-browser")
                .setResultsPerPage(3)
                .setValidReleaseTester(release -> !release.isPreRelease() &&
                        release.getAssets().stream().map(Asset::getName).anyMatch(name -> name.endsWith(".apk")))
                .setCorrectDownloadUrlTester(asset -> asset.getName().endsWith(getFileSuffixForAbi(abi)))
                .build()
                .updateCheck();

        final String version = result.getTagName().replace("iceraven-", "");
        final boolean update = getInstalledVersion(context).map(x -> !x.equals(version)).orElse(true);

        return new UpdateCheckResult.Builder()
                .setUpdateAvailable(update)
                .setDownloadUrl(result.getUrl())
                .setVersion(version)
                .setMetadata(Map.of(FILE_SIZE_BYTES, result.getFileSizeBytes()))
                .build();
    }

    private String getFileSuffixForAbi(ABI abi) {
        switch (abi) {
            case AARCH64:
                return "browser-arm64-v8a-forkRelease.apk";
            case ARM:
                return "browser-armeabi-v7a-forkRelease.apk";
            case X86:
                return "browser-x86-forkRelease.apk";
            case X86_64:
                return "browser-x86_64-forkRelease.apk";
            default:
                throw new ParamRuntimeException("Unknown ABI %s - switch fallthrough", abi);
        }
    }

    @Override
    public void installationCallback(Context context, String installedVersion) {
    }
}

File: ffupdater/src/main/java/de/marmaro/krt/ffupdater/app/impl/Iceraven.kt
package de.marmaro.krt.ffupdater.app.impl

import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.annotation.Keep
import androidx.annotation.MainThread
import de.marmaro.krt.ffupdater.R
import de.marmaro.krt.ffupdater.app.App
import de.marmaro.krt.ffupdater.app.entity.DisplayCategory.BASED_ON_FIREFOX
import de.marmaro.krt.ffupdater.app.entity.LatestVersion
import de.marmaro.krt.ffupdater.device.ABI
import de.marmaro.krt.ffupdater.device.DeviceAbiExtractor
import de.marmaro.krt.ffupdater.network.exceptions.NetworkException
import de.marmaro.krt.ffupdater.network.file.CacheBehaviour
import de.marmaro.krt.ffupdater.network.github.GithubConsumer
import de.marmaro.krt.ffupdater.settings.DeviceSettingsHelper

/**
 * https://github.com/fork-maintainers/iceraven-browser
 * https://api.github.com/repos/fork-maintainers/iceraven-browser/releases
 */
@Keep
object Iceraven : AppBase() {
    override val app = App.ICERAVEN
    override val packageName = "io.github.forkmaintainers.iceraven"
    override val title = R.string.iceraven__title
    override val description = R.string.iceraven__description
    override val installationWarning = R.string.iceraven__warning
    override val downloadSource = "GitHub"
    override val icon = R.drawable.ic_logo_iceraven
    override val minApiLevel = Build.VERSION_CODES.LOLLIPOP
    override val supportedAbis = ARM32_ARM64_X86_X64
    override val signatureHash = "9c0d22379f487b70a4f9f8bec0173cf91a1644f08f93385b5b782ce37660ba81"
    override val projectPage = "https://github.com/fork-maintainers/iceraven-browser"
    override val displayCategory = listOf(BASED_ON_FIREFOX)

    override suspend fun getInstalledVersion(packageManager: PackageManager): String? {
        val installedVersion = super.getInstalledVersion(packageManager)
        return installedVersion?.replace("iceraven-", "")
    }

    @MainThread
    @Throws(NetworkException::class)
    override suspend fun fetchLatestUpdate(context: Context, cacheBehaviour: CacheBehaviour): LatestVersion {
        val fileSuffix = findFileSuffix()
        val result = GithubConsumer.findLatestRelease(
            repository = GithubConsumer.GithubRepo("fork-maintainers", "iceraven-browser", 0),
            isValidRelease = { !it.isPreRelease },
            isSuitableAsset = { it.name.endsWith(fileSuffix) },
            cacheBehaviour = cacheBehaviour,
            requireReleaseDescription = false,
        )
        return LatestVersion(
            downloadUrl = result.url,
            version = result.tagName.replace("iceraven-", ""),
            publishDate = result.releaseDate,
            exactFileSizeBytesOfDownload = result.fileSizeBytes,
            fileHash = null,
        )
    }

    private fun findFileSuffix(): String {
        val fileSuffix = when (DeviceAbiExtractor.findBestAbi(supportedAbis, DeviceSettingsHelper.prefer32BitApks)) {
            ABI.ARMEABI_V7A -> "browser-armeabi-v7a-forkRelease.apk"
            ABI.ARM64_V8A -> "browser-arm64-v8a-forkRelease.apk"
            ABI.X86 -> "browser-x86-forkRelease.apk"
            ABI.X86_64 -> "browser-x86_64-forkRelease.apk"
            else -> throw IllegalArgumentException("ABI is not supported")
        }
        return fileSuffix
    }
}